library/src/main/scala-new-collections/scala/scalajs/js/ArrayOps.scala

```diff
@@ -8,97 +8,2002 @@
+   *    an array consisting of the elements of this array sorted according to
+   *    the ordering `ord`.
+   */
+  def sorted[B >: A](implicit ord: Ordering[B]): js.Array[A] =
+    new js.WrappedArray(xs).sorted(ord.asInstanceOf[Ordering[A]]).array
+
+  /** Sorts this array according to a comparison function.
+   *
+   *  The sort is stable. That is, elements that are equal (as determined by
+   *  `lt`) appear in the same order in the sorted sequence as in the original.
+   *
+   *  @param lt
+   *    the comparison function which tests whether its first argument precedes
+   *    its second argument in the desired ordering.
+   *  @return
+   *    an array consisting of the elements of this array sorted according to
+   *    the comparison function `lt`.
+   */
+  def sortWith(lt: (A, A) => Boolean): js.Array[A] =
+    sorted(Ordering.fromLessThan(lt))
+
+  /** Sorts this array according to the Ordering which results from
+   *  transforming an implicitly given Ordering with a transformation function.
+   *
+   *  @see [[scala.math.Ordering]]
+   *
+   *  @param f
+   *    the transformation function mapping elements to some other domain `B`.
+   *  @param ord
+   *    the ordering assumed on domain `B`.
+   *  @tparam B
+   *    the target type of the transformation `f`, and the type where the
+   *    ordering `ord` is defined.
+   *  @return
+   *    an array consisting of the elements of this array sorted according to
+   *    the ordering where `x < y` if `ord.lt(f(x), f(y))`.
+   */
+  def sortBy[B](f: A => B)(implicit ord: Ordering[B]): js.Array[A] =
+    sorted(ord.on(f))
+
+  /** Creates a non-strict filter of this array.
+   *
+   *  Note: the difference between `c.filter(p)` and `c.withFilter(p)` is that
+   *  the former creates a new array, whereas the latter only restricts the
+   *  domain of subsequent `map`, `flatMap`, `foreach`, and `withFilter`
+   *  operations.
+   *
+   *  @param p
+   *    the predicate used to test elements.
+   *  @return
+   *    an object of class `js.ArrayOps.WithFilter`, which supports `map`,
+   *    `flatMap`, `foreach`, and `withFilter` operations. All these operations
+   *    apply to those elements of this array which satisfy the predicate `p`.
+   */
+  def withFilter(p: A => Boolean): js.ArrayOps.WithFilter[A] =
+    new js.ArrayOps.WithFilter[A](p, xs)
+
+  /** Finds the index of the first occurrence of some value in this array after
+   *  or at some start index.
+   *
+   *  @param elem
+   *    the element value to search for.
+   *  @param from
+   *    the start index
+   *  @return
+   *    the index `>= from` of the first element of this array that is equal
+   *    (as determined by `==`) to `elem`, or `-1`, if none exists.
+   */
+  def indexOf(elem: A, from: Int = 0): Int = {
+    // scalastyle:off return
+    val len = xs.length
+    var i = from
+    while (i < len) {
+      if (elem == xs(i))
+        return i
+      i += 1
+    }
+    -1
+    // scalastyle:on return
```

Comment:
From my mails: @gzm0 says:

> Why not a tailrec method instead? Should compile to roughly the same code.

Reply:
Because the copied from [`scala.collection.ArrayOps.scala`](https://github.com/scala/scala/blob/008d525314ca55b41652c79fb23ba1c9cf78a133/src/library/scala/collection/ArrayOps.scala#L499-L506).

In general, virtually all the methods are copied from their implementation in upstream scala/scala.